### PR TITLE
[gui] Add some built-int math APIs for building translation, scale and rotation matrix

### PR DESCRIPTION
**Pull request by Morcki**
Related issue = #5513 
Related issue = #5735 

**Usage**
```py
import taichi as ti
import math
# Get translation matrix
translate_vec = ti.math.vec3(1., 2., 3.)
translate_mat = ti.math.translate(translate_vec[0], translate_vec[1], translate_vec[2])
# Get scale matrix
scale_vec = ti.math.vec3(1., 2., 3.)
scale_mat = ti.math.scale(scale_vec[0], scale_vec[1], scale_vec[2])
# Get rotation matrix
angleZ =  1.046
angleX = 0.52
angleY = -0.785
rotation_mat = ti.math.rotation3d(angleX, angleY, angleZ)
# or call as blew, which is the same restult with rotation3d
# rotation_mat = ti.math.rot_yaw_pitch_roll(angleZ, angleX, angleY)

# Then build a transformation matrix
trans_mat = translate_mat @ rotation_mat @ scale_mat

# Additional
# usage of rot_by_axis
Angle = math.pi * 0.5
X = ti.math.vec3(1.0, 0.0, 0.0)
Y = ti.math.vec3(0.0, 1.0, 0.0, 1.0)
# rotate Y by an angle with X axis
Y1 = ti.math.rot_by_axis(Angle, X) @ Y
```

**Mainly new built-in APIs**
Details please refer to the notes of each API
Note that all the rotations ars based on counter-clockwise

- Translation
  - ti.math.translate(dx, dy, dz)
- Scale
  - ti.math.scale(sx, sy, sz)
- Rotation
  - ti.math.rot_by_axis(ang, axis)
  - ti.math.rot_yaw_pitch_roll(yaw, pitch, roll)
  - ti.math.rotation3d(ang_x, ang_y, ang_z)

**Old version deprecated**

These APIs not used by python users, so directly drop these APIs.

- ti.math.rot2d -> moved to ti.math.rotation2d
- ti.math.rot3d
- ti.math.rotate2d
- ti.math.rotate3d

Add deprecated warning:
- ti.Matrix.rotation2d -> will be moved to ti.math.rotation2d

### CODE REVIEW

**Comment by neozhaoliang** [comment]
I think we should remove the two functions `rot3d`, `rotate3d` in `math` module, as well as the classmethod `Matrix.rotation2d`?

**Comment by Morcki** [comment]
> I think we should remove the two functions `rot3d`, `rotate3d` in `math` module, as well as the classmethod `Matrix.rotation2d`?

Add deprecated warning : `Matrix.rotation2d`
Removed : `rot2d`, `rot3d`, `rotate2d`, `rotate3d`
New add : `rotation2d`

File: python/taichi/math/mathimpl.py [hunk 1]
```diff
@@ -673,40 +588,40 @@ def rot2(ang):
...
+        :class:`~taichi.math.mat4`: rotation matrix
+    Example:
+        >>> import math
+        >>> rotation3d(0.52, -0.785, 1.046)
+        [[ 0.05048351 -0.61339645 -0.78816002  0.        ]
+        [ 0.65833154  0.61388511 -0.4355969   0.        ]
+        [ 0.75103329 -0.49688014  0.4348093   0.        ]
+        [ 0.          0.          0.          1.        ]]
+    """
+    return rot3d_yaw_pitch_roll(ang_z, ang_x, ang_y)
```
Comment (neozhaoliang) [inline python/taichi/math/mathimpl.py:617]:
```suggestion

```
Should be two blank lines here

Comment (Morcki) [reply python/taichi/math/mathimpl.py:617]: Thanks

File: python/taichi/math/mathimpl.py [hunk 2]
```diff
@@ -469,71 +469,111 @@ def mod(x, y):
...
+         [ 0 0 3 0]
+         [ 0 0 0 1]]
+    """
+    return mat4([[sx, 0., 0., 0.], [0., sy, 0., 0.], [0., 0., sz, 0.],
+                 [0., 0., 0., 1.]])
+
+
+@ti.func
+def rot3d_by_axis(ang, axis):
+    """rotate the matrix by an angle with the vector as the rotation axis
```
Comment (neozhaoliang) [inline python/taichi/math/mathimpl.py:523]:
```suggestion
def rot3d_by_axis(axis, ang):
    """Returns the 3x3 matrix representation of the rotation with given axis `axis` and angle `ang`. The vector `axis` must be normalized.
```

Comment (neozhaoliang) [reply python/taichi/math/mathimpl.py:523]: Quite confusing the name `rot3d_xxx` returns a 4x4 matrix here.

File: python/taichi/math/mathimpl.py [hunk 3]
```diff
@@ -469,71 +469,111 @@ def mod(x, y):
...
+                     temp[2] * axis[0] + s * axis[1],
+                     temp[2] * axis[1] - s * axis[0], c + temp[2] * axis[2], 0.
+                 ], [0., 0., 0., 1.]])
 
 
 @ti.func
-def eye(n: ti.template()):
-    """Returns the nxn identity matrix.
+def rot3d_yaw_pitch_roll(yaw, pitch, roll):
+    """Creates a 3D 4 * 4 homogeneous rotation matrix from euler angles(Y * X * Z).
```
Comment (neozhaoliang) [inline python/taichi/math/mathimpl.py:553]:
```suggestion
    """Returns a 4x4 homogeneous rotation matrix representing the 3d rotation with Euler angles (yaw, pitch, roll).
```

Comment (neozhaoliang) [reply python/taichi/math/mathimpl.py:553]: we should also explain what are the axes of the three rotations and in which order they are applied.

File: python/taichi/math/mathimpl.py [hunk 4]
```diff
@@ -548,40 +588,41 @@ def rot2(ang):
...
     ca, sa = ti.cos(ang), ti.sin(ang)
     return mat2([[ca, -sa], [sa, ca]])
 
 
 @ti.func
-def rot3(axis, ang):
-    """Returns the matrix representation of a 3d rotation,
-    given the axis and angle of rotation.
+def rotation3d(ang_x, ang_y, ang_z):
+    """ Creates a 3D 4 * 4 homogeneous rotation matrix from an euler angle(Y * X * Z).
```
Comment (neozhaoliang) [inline python/taichi/math/mathimpl.py:600]: The same as above, we should explain more on the axis and the order of rotations.

File: tests/python/test_math_module.py [hunk 5]
```diff
@@ -79,3 +91,78 @@ def test():
...
+    axisZ = ti.math.vec3(0.0, 0.0, 1.0)
+
+    rotationEuler = ti.math.rot3d_yaw_pitch_roll(first, second, third)
+    rotationInvertedY = ti.math.rot3d_by_axis(
+        third, axisZ) @ ti.math.rot3d_by_axis(
+            second, axisX) @ ti.math.rot3d_by_axis(-first, axisY)
+    rotationDumb = ti.Matrix.zero(ti.f32, 4, 4)
+    rotationDumb = ti.math.rot3d_by_axis(first, axisY) @ rotationDumb
+    rotationDumb = ti.math.rot3d_by_axis(second, axisX) @ rotationDumb
+    rotationDumb = ti.math.rot3d_by_axis(third, axisZ) @ rotationDumb
```
Comment (neozhaoliang) [inline tests/python/test_math_module.py:150]:
```suggestion
    rotationInvertedY = ti.math.rot3d_by_axis(
        axisZ, third) @ ti.math.rot3d_by_axis(
            axisX, second) @ ti.math.rot3d_by_axis(axisY, -first)
    rotationDumb = ti.Matrix.zero(ti.f32, 4, 4)
    rotationDumb = ti.math.rot3d_by_axis(axisY, first) @ rotationDumb
    rotationDumb = ti.math.rot3d_by_axis(axisX, second) @ rotationDumb
    rotationDumb = ti.math.rot3d_by_axis(axisZ, third) @ rotationDumb
```

File: python/taichi/math/mathimpl.py (hunk 3 above)
Comment (Morcki) [reply python/taichi/math/mathimpl.py:553]: I think we should indicate how we rotate with what order and what axis.

```py
Returns a 4x4 homogeneous rotation matrix representing
the 3d rotation with Euler angles (rotate with Y axis fitst, X axis second, Z axis third).
```

Comment (neozhaoliang) [reply python/taichi/math/mathimpl.py:553]: This should be fine.

File: python/taichi/math/mathimpl.py (hunk 2 above)
Comment (Morcki) [reply python/taichi/math/mathimpl.py:523]: But that's our final name, maybe here, the meaning of rot3d_xxx is rotation in 3D space, but the 4x4 matrix means a affine transformation

Comment (neozhaoliang) [reply python/taichi/math/mathimpl.py:523]: Shall we hide  "3d" in the name and just call it `rot_by_axis`?

Comment (Morcki) [reply python/taichi/math/mathimpl.py:523]: But there are still `rotation2d`, `rotation3d`, which need us to change their names if so.

**Review by neozhaoliang — APPROVED** [review]
LGTM! And thanks for the jobs!

**Pull request merged**